selene_sdk/samplers/random_positions_sampler_original.py

```diff
@@ -0,0 +1,331 @@
+        would like to predict.
+    validation_holdout : list(str) or float
+        The samples to hold out for validating model performance. These
+        can be "regional" or "proportional". If regional, this is a list
+        of region names (e.g. `['chrX', 'chrY']`). These regions must
+        match those specified in the first column of the tabix-indexed
+        BED file. If proportional, this is the fraction of total samples
+        that will be held out.
+    test_holdout : list(str) or float
+        The samples to hold out for testing model performance. See the
+        documentation for `validation_holdout` for more details.
+    sequence_length : int
+        The length of the sequences to  train the model on.
+    bin_radius : int
+        From the center of the sequence, the radius in which to detect
+        a feature annotation in order to include it as a sample's label.
+    surrounding_sequence_radius : int
+        The length of sequence falling outside of the feature detection
+        bin (i.e. `bin_radius`) center, but still within the
+        `sequence_length`.
+    modes : list(str)
+        The list of modes that the sampler can be run in.
+    mode : str
+        The current mode that the sampler is running in. Must be one of
+        the modes listed in `modes`.
+    """
+    def __init__(self,
+                 reference_sequence,
+                 target_path,
+                 features,
+                 seed=436,
+                 validation_holdout=['chr6', 'chr7'],
+                 test_holdout=['chr8', 'chr9'],
+                 sequence_length=1000,
+                 center_bin_to_predict=200,
+                 feature_thresholds=0.5,
+                 mode="train",
+                 save_datasets=[],
+                 output_dir=None):
+        super(RandomPositionsSampler, self).__init__(
+            reference_sequence,
+            target_path,
+            features,
+            seed=seed,
+            validation_holdout=validation_holdout,
+            test_holdout=test_holdout,
+            sequence_length=sequence_length,
+            center_bin_to_predict=center_bin_to_predict,
+            feature_thresholds=feature_thresholds,
+            mode=mode,
+            save_datasets=save_datasets,
+            output_dir=output_dir)
+
+        self._sample_from_mode = {}
+        self._randcache = {}
+        for mode in self.modes:
+            self._sample_from_mode[mode] = None
+            self._randcache[mode] = {"cache_indices": None, "sample_next": 0}
+
+        self.sample_from_intervals = []
+        self.interval_lengths = []
+
+        if self._holdout_type == "chromosome":
+            self._partition_genome_by_chromosome()
+        else:
+            self._partition_genome_by_proportion()
+
+        for mode in self.modes:
+            self._update_randcache(mode=mode)
+
+    def _partition_genome_by_proportion(self):
```

Comment:
You'll notice that now that `_partition_by_proportion` is going to store VALID POSITIONS from which to sample. this dictionary of positions should be kept if user doesn't specify N_train, N_validation and (optionally) N_test, so that the user can SAMPLE WITH REPLACEMENT from the same positions for each partition. This means we might need the randcache to be updated differently for `_partition_by_chromosome` compared to `_partition_by_proportion` - I don't know yet! I think the `randcache` should maintain the same data structure it has currently: dictionary of mode -> list of positions, but the way it is updated might be different depending on how we handle partition by chrom vs proportion.